### ISSUE TITLE
Theoretical integer overflow opportunity in poolGrow

### ISSUE THREAD

**Issue by RMJ10** [libexpat/libexpat#25]
The string pool code for allocating a bigger pool, poolGrow(), has a potential issue when attempting to reallocate memory.  The function starts off well enough (xmlparse.c lines 6325-6330):
```
    BLOCK *temp;
    int blockSize = (int)((unsigned)(pool->end - pool->start)*2U);

    if (blockSize < 0)
      return XML_FALSE;
```
This avoids any issue with signed integer overflow when doubling the current size of the string pool and bails out if ludicrous amounts of memory are needed.  Unfortunately this isn't the whole story (line 6331-6337):
```
    temp = (BLOCK *)
      pool->mem->realloc_fcn(pool->blocks,
                             (offsetof(BLOCK, s)
                              + blockSize * sizeof(XML_Char)));
    if (temp == NULL)
      return XML_FALSE;
```
`sizeof(XML_Char)` will be 1 for most builds, but if `XML_UNICODE` is defined it will be 2, since characters will be held as UTF-16 internally.  So we could be doubling the allocation request and adding a bit more for good measure, which obviously has the potential to overflow.

(In reality I would expect the allocator to have failed long before this code would get called.  It does require us to have allocated and run out of an enormous amount of memory first.  However the moment anyone says it's impossible, someone will manage to do it :-)

**Comment by hartwork** [comment]
`sizeof` yields unsigned `size_t` so the multiplication is unsigned, i.e. no undefined behavior (but still risk of overflow). I'm removing the "signed" part from the bug title therefore.

**Comment by hartwork** [comment]
Should be fixed by #26, closing.
